Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_usb_wfssrv.cpp

```diff
@@ -0,0 +1,266 @@
+  u32 buffer_in = Memory::Read_U32(command_address + 0x10);
+  u32 buffer_in_size = Memory::Read_U32(command_address + 0x14);
+  u32 buffer_out = Memory::Read_U32(command_address + 0x18);
+  u32 buffer_out_size = Memory::Read_U32(command_address + 0x1C);
+
+  int return_error_code = FS_SUCCESS;
+
+  switch (command)
+  {
+  case IOCTL_WFS_INIT:
+    // TODO(wfs): Implement.
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_INIT");
+    break;
+
+  case IOCTL_WFS_DEVICE_INFO:
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_DEVICE_INFO");
+    Memory::Write_U64(16ull << 30, buffer_out);  // 16GB storage.
+    Memory::Write_U8(4, buffer_out + 8);
+    break;
+
+  case IOCTL_WFS_GET_DEVICE_NAME:
+  {
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_GET_DEVICE_NAME");
+    Memory::Write_U8(static_cast<u8>(m_device_name.size()), buffer_out);
+    Memory::CopyToEmu(buffer_out + 1, m_device_name.data(), m_device_name.size());
+    break;
+  }
+
+  case IOCTL_WFS_ATTACH_DETACH_2:
+    // TODO(wfs): Implement.
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_ATTACH_DETACH_2(%d)", command);
+    Memory::Write_U32(1, buffer_out);
+    Memory::Write_U32(0, buffer_out + 4);  // device id?
+    Memory::Write_U32(0, buffer_out + 8);
+    break;
+
+  case IOCTL_WFS_ATTACH_DETACH:
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_ATTACH_DETACH(%d)", command);
+    Memory::Write_U32(1, buffer_out);
+    Memory::Write_U32(0, buffer_out + 4);
+    Memory::Write_U32(0, buffer_out + 8);
+    return GetNoReply();
+
+  // TODO(wfs): Globbing is not really implemented, we just fake the one case
+  // (listing /vol/*) which is required to get the installer to work.
+  case IOCTL_WFS_GLOB_START:
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_GLOB_START(%d)", command);
+    Memory::Memset(buffer_out, 0, buffer_out_size);
+    memcpy(Memory::GetPointer(buffer_out + 0x14), m_device_name.data(), m_device_name.size());
+    break;
+
+  case IOCTL_WFS_GLOB_NEXT:
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_GLOB_NEXT(%d)", command);
+    return_error_code = WFS_EEMPTY;
+    break;
+
+  case IOCTL_WFS_GLOB_END:
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_GLOB_END(%d)", command);
+    Memory::Memset(buffer_out, 0, buffer_out_size);
+    break;
+
+  case IOCTL_WFS_OPEN:
+  {
+    u32 mode = Memory::Read_U32(buffer_in);
+    u16 path_len = Memory::Read_U16(buffer_in + 0x20);
+    std::string path = Memory::GetString(buffer_in + 0x22, path_len);
+
+    u16 fd = GetNewFileDescriptor();
+    FileDescriptor* fd_obj = &m_fds[fd];
+    fd_obj->in_use = true;
+    fd_obj->path = path;
+    fd_obj->mode = mode;
+    fd_obj->position = 0;
+
+    if (!fd_obj->Open())
+    {
+      ERROR_LOG(WII_IPC_HLE, "IOCTL_WFS_OPEN(%s, %d): error opening file", path.c_str(), mode);
+      ReleaseFileDescriptor(fd);
+      return_error_code = -1;  // TODO(wfs): proper error code.
+      break;
+    }
+
+    INFO_LOG(WII_IPC_HLE, "IOCTL_WFS_OPEN(%s, %d) -> %d", path.c_str(), mode, fd);
+    Memory::Write_U16(fd, buffer_out + 0x14);
```

Comment:
See that there are a few offsets that are used multiple times.  For the sake of clarity, does it make sense to define them in an anonymous namespace so as to call them out (maybe with a comment as to what they are if the variable name isn't sufficient).